src/vset.c

```diff
@@ -0,0 +1,1249 @@
+            long long max_expiry = getExpiry(pvGet(vec, len - 1));
+            if (get_max_bucket_ts(min_expiry) == get_max_bucket_ts(max_expiry)) {
+                /* In case we can just insert the bucket, no need to iterate and insert it's elements. we can just push the bucket as a whole. */
+                unsigned char key[VSET_BUCKET_KEY_LEN] = {0};
+                size_t key_len = encodeNewExpiryBucketKey(key, max_expiry);
+                raxInsert(r, key, key_len, expiry_buckets, NULL);
+                expiry_buckets = vsetBucketSetRax(expiry_buckets, r);
+                expiry_buckets = insertToBucket_RAX(getExpiry, expiry_buckets, entry, expiry);
+            } else {
+                /* We need to migrate entries to the new set of buckets since we do not know all entries are in the same bucket */
+                expiry_buckets = vsetBucketSetRax(expiry_buckets, r);
+                for (uint32_t i = 0; i < len; i++) {
+                    void *moved_entry = pvGet(vec, i);
+                    expiry_buckets = insertToBucket_RAX(getExpiry, expiry_buckets, moved_entry, getExpiry(moved_entry));
+                }
+                /* free the vector */
+                pvFree(vec);
+                /* now insert the new entry to the buckets */
+                expiry_buckets = insertToBucket_RAX(getExpiry, expiry_buckets, entry, expiry);
+            }
+        } else {
+            expiry_buckets = insertToBucket_VECTOR(getExpiry, expiry_buckets, entry, expiry);
+        }
+        break;
+    }
+    case VSET_BUCKET_RAX:
+        expiry_buckets = insertToBucket_RAX(getExpiry, expiry_buckets, entry, expiry);
+        break;
+    default:
+        panic("Cannot insert to bucket which is not single, vector or rax");
+    }
+    /* update the set */
+    *set = expiry_buckets;
+    return 1;
+}
+
+int vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry, long long expiry) {
+    bool removed;
+    vsetBucket *bucket = *set;
+    int bucket_type = vsetBucketType(bucket);
+    switch (bucket_type) {
+    case VSET_BUCKET_NONE:
+        /* We cannot remove from empty set */
+        return 0;
+    case VSET_BUCKET_SINGLE:
+        bucket = removeFromBucket_SINGLE(getExpiry, bucket, entry, expiry, &removed);
+        break;
+    case VSET_BUCKET_VECTOR:
+        bucket = removeFromBucket_VECTOR(getExpiry, bucket, entry, expiry, &removed);
+        break;
+    case VSET_BUCKET_HT:
+        bucket = removeFromBucket_HASHTABLE(getExpiry, bucket, entry, expiry, &removed);
+        break;
+    case VSET_BUCKET_RAX:
+        bucket = removeFromBucket_RAX(getExpiry, bucket, entry, expiry, &removed);
+        break;
+    default:
+        panic("Cannot insert to bucket which is not single, vector or rax");
+    }
+    *set = bucket;
+    return removed ? 1 : 0;
+}
+
+int vsetUpdateEntry(vset *set, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry) {
```

Comment:
The expiry timestamps are passed explicitly to this function just as an optimization? It seems redundant because this function could just call getExpiry on them. It would be a simpler API but maybe inefficient?

Btw, a top comment for the public functions (1-2 lines) could be useful...

Reply:
I fixed so that only update will get explicit expiration time. I think that is indeed a much better API.